### ISSUE TITLE
Callable with TypeVar input doesn't typecheck properly

### ISSUE THREAD

**Issue by Kache** [microsoft/pyright#6133]
**Describe the bug**
A `fun: Callable[[T], T]` doesn't typecheck properly.

**Code or Screenshots**

```python
from typing import Callable, TypeVar

T = TypeVar('T')

def bar(fun: Callable[[T], T]):
    fun(123)

bar(lambda x: x)
```

My actual use-case is more like:

```python
def foo(val: T, extract: Callable[[Deque[T]], T]):
    deq = deque([(val, [val])])
    tup = extract(deq)  # e.g. extract in [deque.pop, deque.popleft]
    return tup[0]
```

**VS Code extension or command-line**
coc-pyright `[coc.nvim] coc-pyright 1.1.325 with Pyright 1.1.326`, but also repro'd with CLI `pyright 1.1.331`:

```sh
❯ pyright generic_callable.py
/path/generic_callable.py
  /path/generic_callable.py:6:9 - error: Argument of type "Literal[123]" cannot be assigned to parameter of type "T@bar"
    Type "Literal[123]" cannot be assigned to type "T@bar" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
```

**Comment by erictraut** [comment]
Pyright's behavior is correct here. The type variable `T` is scoped to the function `bar`, not to the callable passed to `bar`. If your intent is to scope `T` to the callable, you need to use a callback protocol.

```python
from typing import Protocol, TypeVar

T = TypeVar("T")

class CallabackProto(Protocol):
    def __call__(self, x: T) -> T:
        ...

def bar(fun: CallabackProto):
    fun(123)

bar(lambda x: x)
```
